File: typescript/factory/framework/Factory.ts
import { Product } from './Product.ts';

export abstract class Factory {
  protected abstract createdProduct(owner: string): Product;
  protected abstract registerProduct(product: Product): void;

  public create = (owner: string): Product => {
    const p: Product = this.createdProduct(owner);
    this.registerProduct(p);
    return p;
  };
}

File: go/factory/framework/factory.go
package framework

type Factory interface {
	CreateProduct(owner string) Product
	RegisterProduct(product Product)
}

type AbstractFactory struct {
	Factory Factory
}

func (af *AbstractFactory) Create(owner string) Product {
	p := af.Factory.CreateProduct(owner)
	af.Factory.RegisterProduct(p)
	return p
}
